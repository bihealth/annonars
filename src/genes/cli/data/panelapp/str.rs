/* 
 * PanelApp API
 *
 * PanelApp API
 *
 * OpenAPI spec version: v1
 * Contact: panelapp@genomicsengland.co.uk
 * Generated by: https://github.com/swagger-api/swagger-codegen.git
 */


#[allow(unused_imports)]
use serde_json::Value;

#[derive(Debug, Serialize, Deserialize)]
pub struct Str {
  #[serde(rename = "gene_data")]
  gene_data: Option<Value>,
  #[serde(rename = "entity_type")]
  entity_type: String,
  #[serde(rename = "entity_name")]
  entity_name: String,
  #[serde(rename = "confidence_level")]
  confidence_level: String,
  #[serde(rename = "penetrance")]
  penetrance: Option<String>,
  #[serde(rename = "publications")]
  publications: Vec<String>,
  #[serde(rename = "evidence")]
  evidence: Vec<String>,
  #[serde(rename = "phenotypes")]
  phenotypes: Vec<String>,
  #[serde(rename = "mode_of_inheritance")]
  mode_of_inheritance: String,
  #[serde(rename = "repeated_sequence")]
  repeated_sequence: String,
  #[serde(rename = "chromosome")]
  chromosome: String,
  #[serde(rename = "grch37_coordinates")]
  grch37_coordinates: Vec<i32>,
  #[serde(rename = "grch38_coordinates")]
  grch38_coordinates: Vec<i32>,
  /// =< Maximum normal number of repeats
  #[serde(rename = "normal_repeats")]
  normal_repeats: i32,
  /// >= Minimum fully penetrant pathogenic number of repeats
  #[serde(rename = "pathogenic_repeats")]
  pathogenic_repeats: i32,
  #[serde(rename = "tags")]
  tags: Option<Vec<String>>,
  #[serde(rename = "panel")]
  panel: Option<::models::Panel>
}

impl Str {
  pub fn new(entity_type: String, entity_name: String, confidence_level: String, publications: Vec<String>, evidence: Vec<String>, phenotypes: Vec<String>, mode_of_inheritance: String, repeated_sequence: String, chromosome: String, grch37_coordinates: Vec<i32>, grch38_coordinates: Vec<i32>, normal_repeats: i32, pathogenic_repeats: i32) -> Str {
    Str {
      gene_data: None,
      entity_type: entity_type,
      entity_name: entity_name,
      confidence_level: confidence_level,
      penetrance: None,
      publications: publications,
      evidence: evidence,
      phenotypes: phenotypes,
      mode_of_inheritance: mode_of_inheritance,
      repeated_sequence: repeated_sequence,
      chromosome: chromosome,
      grch37_coordinates: grch37_coordinates,
      grch38_coordinates: grch38_coordinates,
      normal_repeats: normal_repeats,
      pathogenic_repeats: pathogenic_repeats,
      tags: None,
      panel: None
    }
  }

  pub fn set_gene_data(&mut self, gene_data: Value) {
    self.gene_data = Some(gene_data);
  }

  pub fn with_gene_data(mut self, gene_data: Value) -> Str {
    self.gene_data = Some(gene_data);
    self
  }

  pub fn gene_data(&self) -> Option<&Value> {
    self.gene_data.as_ref()
  }

  pub fn reset_gene_data(&mut self) {
    self.gene_data = None;
  }

  pub fn set_entity_type(&mut self, entity_type: String) {
    self.entity_type = entity_type;
  }

  pub fn with_entity_type(mut self, entity_type: String) -> Str {
    self.entity_type = entity_type;
    self
  }

  pub fn entity_type(&self) -> &String {
    &self.entity_type
  }


  pub fn set_entity_name(&mut self, entity_name: String) {
    self.entity_name = entity_name;
  }

  pub fn with_entity_name(mut self, entity_name: String) -> Str {
    self.entity_name = entity_name;
    self
  }

  pub fn entity_name(&self) -> &String {
    &self.entity_name
  }


  pub fn set_confidence_level(&mut self, confidence_level: String) {
    self.confidence_level = confidence_level;
  }

  pub fn with_confidence_level(mut self, confidence_level: String) -> Str {
    self.confidence_level = confidence_level;
    self
  }

  pub fn confidence_level(&self) -> &String {
    &self.confidence_level
  }


  pub fn set_penetrance(&mut self, penetrance: String) {
    self.penetrance = Some(penetrance);
  }

  pub fn with_penetrance(mut self, penetrance: String) -> Str {
    self.penetrance = Some(penetrance);
    self
  }

  pub fn penetrance(&self) -> Option<&String> {
    self.penetrance.as_ref()
  }

  pub fn reset_penetrance(&mut self) {
    self.penetrance = None;
  }

  pub fn set_publications(&mut self, publications: Vec<String>) {
    self.publications = publications;
  }

  pub fn with_publications(mut self, publications: Vec<String>) -> Str {
    self.publications = publications;
    self
  }

  pub fn publications(&self) -> &Vec<String> {
    &self.publications
  }


  pub fn set_evidence(&mut self, evidence: Vec<String>) {
    self.evidence = evidence;
  }

  pub fn with_evidence(mut self, evidence: Vec<String>) -> Str {
    self.evidence = evidence;
    self
  }

  pub fn evidence(&self) -> &Vec<String> {
    &self.evidence
  }


  pub fn set_phenotypes(&mut self, phenotypes: Vec<String>) {
    self.phenotypes = phenotypes;
  }

  pub fn with_phenotypes(mut self, phenotypes: Vec<String>) -> Str {
    self.phenotypes = phenotypes;
    self
  }

  pub fn phenotypes(&self) -> &Vec<String> {
    &self.phenotypes
  }


  pub fn set_mode_of_inheritance(&mut self, mode_of_inheritance: String) {
    self.mode_of_inheritance = mode_of_inheritance;
  }

  pub fn with_mode_of_inheritance(mut self, mode_of_inheritance: String) -> Str {
    self.mode_of_inheritance = mode_of_inheritance;
    self
  }

  pub fn mode_of_inheritance(&self) -> &String {
    &self.mode_of_inheritance
  }


  pub fn set_repeated_sequence(&mut self, repeated_sequence: String) {
    self.repeated_sequence = repeated_sequence;
  }

  pub fn with_repeated_sequence(mut self, repeated_sequence: String) -> Str {
    self.repeated_sequence = repeated_sequence;
    self
  }

  pub fn repeated_sequence(&self) -> &String {
    &self.repeated_sequence
  }


  pub fn set_chromosome(&mut self, chromosome: String) {
    self.chromosome = chromosome;
  }

  pub fn with_chromosome(mut self, chromosome: String) -> Str {
    self.chromosome = chromosome;
    self
  }

  pub fn chromosome(&self) -> &String {
    &self.chromosome
  }


  pub fn set_grch37_coordinates(&mut self, grch37_coordinates: Vec<i32>) {
    self.grch37_coordinates = grch37_coordinates;
  }

  pub fn with_grch37_coordinates(mut self, grch37_coordinates: Vec<i32>) -> Str {
    self.grch37_coordinates = grch37_coordinates;
    self
  }

  pub fn grch37_coordinates(&self) -> &Vec<i32> {
    &self.grch37_coordinates
  }


  pub fn set_grch38_coordinates(&mut self, grch38_coordinates: Vec<i32>) {
    self.grch38_coordinates = grch38_coordinates;
  }

  pub fn with_grch38_coordinates(mut self, grch38_coordinates: Vec<i32>) -> Str {
    self.grch38_coordinates = grch38_coordinates;
    self
  }

  pub fn grch38_coordinates(&self) -> &Vec<i32> {
    &self.grch38_coordinates
  }


  pub fn set_normal_repeats(&mut self, normal_repeats: i32) {
    self.normal_repeats = normal_repeats;
  }

  pub fn with_normal_repeats(mut self, normal_repeats: i32) -> Str {
    self.normal_repeats = normal_repeats;
    self
  }

  pub fn normal_repeats(&self) -> &i32 {
    &self.normal_repeats
  }


  pub fn set_pathogenic_repeats(&mut self, pathogenic_repeats: i32) {
    self.pathogenic_repeats = pathogenic_repeats;
  }

  pub fn with_pathogenic_repeats(mut self, pathogenic_repeats: i32) -> Str {
    self.pathogenic_repeats = pathogenic_repeats;
    self
  }

  pub fn pathogenic_repeats(&self) -> &i32 {
    &self.pathogenic_repeats
  }


  pub fn set_tags(&mut self, tags: Vec<String>) {
    self.tags = Some(tags);
  }

  pub fn with_tags(mut self, tags: Vec<String>) -> Str {
    self.tags = Some(tags);
    self
  }

  pub fn tags(&self) -> Option<&Vec<String>> {
    self.tags.as_ref()
  }

  pub fn reset_tags(&mut self) {
    self.tags = None;
  }

  pub fn set_panel(&mut self, panel: ::models::Panel) {
    self.panel = Some(panel);
  }

  pub fn with_panel(mut self, panel: ::models::Panel) -> Str {
    self.panel = Some(panel);
    self
  }

  pub fn panel(&self) -> Option<&::models::Panel> {
    self.panel.as_ref()
  }

  pub fn reset_panel(&mut self) {
    self.panel = None;
  }

}



