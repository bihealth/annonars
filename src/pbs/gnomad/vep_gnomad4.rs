//! Code generate for protobufs by `prost-build`.

use std::str::FromStr;

include!(concat!(env!("OUT_DIR"), "/annonars.gnomad.vep_gnomad4.rs"));
include!(concat!(
    env!("OUT_DIR"),
    "/annonars.gnomad.vep_gnomad4.serde.rs"
));

impl Vep {
    /// Returns number of fields in a gnomAD v4 VEP entry.
    pub fn num_fields() -> usize {
        46
    }
}

impl FromStr for Vep {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let values = s.split('|').collect::<Vec<_>>();

        if values.len() < 2 {
            dbg!(s);
        }

        Ok(Vep {
            allele: values[0].to_string(),
            consequence: values[1].to_string(),
            impact: values[2].to_string(),
            symbol: values[3].to_string(),
            gene: values[4].to_string(),
            feature_type: values[5].to_string(),
            feature: values[6].to_string(),
            feature_biotype: values[7].to_string(),
            exon: (!values[8].is_empty()).then(|| values[8].to_string()),
            intron: (!values[9].is_empty()).then(|| values[9].to_string()),
            hgvsc: (!values[10].is_empty()).then(|| values[10].to_string()),
            hgvsp: (!values[11].is_empty()).then(|| values[11].to_string()),
            cdna_position: (!values[12].is_empty()).then(|| values[12].to_string()),
            cds_position: (!values[13].is_empty()).then(|| values[13].to_string()),
            protein_position: (!values[14].is_empty()).then(|| values[14].to_string()),
            amino_acids: (!values[15].is_empty()).then(|| values[15].to_string()),
            codons: (!values[16].is_empty()).then(|| values[16].to_string()),
            allele_num: (!values[17].is_empty())
                .then(|| values[17].parse())
                .transpose()
                .map_err(|e| anyhow::anyhow!("problem parsing vep/ALLELE_NUM: {}", e))?,
            distance: (!values[18].is_empty()).then(|| values[18].to_string()),
            strand: (!values[19].is_empty()).then(|| values[19].to_string()),
            flags: (!values[20].is_empty()).then(|| values[20].to_string()),
            variant_class: (!values[21].is_empty()).then(|| values[21].to_string()),
            symbol_source: (!values[22].is_empty()).then(|| values[22].to_string()),
            hgnc_id: (!values[23].is_empty()).then(|| values[23].to_string()),
            canonical: (!values[24].is_empty()).then(|| values[24] == "YES"),
            mane_select: (!values[25].is_empty()).then(|| values[25] == "YES"),
            mane_plus_clinical: (!values[26].is_empty()).then(|| values[26] == "YES"),
            tsl: (!values[27].is_empty())
                .then(|| values[27].parse())
                .transpose()
                .map_err(|e| anyhow::anyhow!("problem parsing vep/TSL: {}", e))?,
            appris: (!values[28].is_empty()).then(|| values[28].to_string()),
            ccds: (!values[29].is_empty()).then(|| values[29].to_string()),
            ensp: (!values[30].is_empty()).then(|| values[30].to_string()),
            uniprot_isoform: (!values[31].is_empty()).then(|| values[31].to_string()),
            source: (!values[32].is_empty()).then(|| values[32].to_string()),
            domains: (!values[33].is_empty())
                .then(|| {
                    let pairs = values[33].split('&').collect::<Vec<_>>();
                    pairs
                        .into_iter()
                        .filter(|s| *s != "null")
                        .map(|p| {
                            let tmp = p.split(':').collect::<Vec<_>>();
                            super::vep_common::Domain {
                                source: tmp[0].to_string(),
                                id: tmp[1].to_string(),
                            }
                        })
                        .collect::<Vec<_>>()
                })
                .unwrap_or_default(),
            mirna: (!values[34].is_empty()).then(|| values[34].to_string()),
            hgvs_offset: (!values[35].is_empty()).then(|| values[35].to_string()),
            pubmed: (!values[36].is_empty()).then(|| values[36].to_string()),
            motif_name: (!values[37].is_empty()).then(|| values[37].to_string()),
            motif_pos: (!values[38].is_empty()).then(|| values[38].to_string()),
            high_inf_pos: (!values[39].is_empty()).then(|| values[39].to_string()),
            motif_score_change: (!values[40].is_empty()).then(|| values[40].to_string()),
            transcription_factors: (!values[41].is_empty()).then(|| values[41].to_string()),
            lof: (!values[42].is_empty()).then(|| values[42].to_string()),
            lof_filter: (!values[43].is_empty()).then(|| values[43].to_string()),
            lof_flags: (!values[44].is_empty()).then(|| values[44].to_string()),
            lof_info: (!values[45].is_empty()).then(|| values[45].to_string()),
        })
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_vep_from_string() {
        let s = "\
        C|upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|\
        Transcript|ENST00000450305|transcribed_unprocessed_pseudogene||||||||||\
        1|1979|1||SNV|HGNC|HGNC:37102|YES||||||||Ensembl|||||||||||||||,C|\
        upstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|Transcript|\
        ENST00000456328|processed_transcript||||||||||1|1838|1||SNV|HGNC|\
        HGNC:37102||||1|||||Ensembl|||||||||||||||,C|downstream_gene_variant|\
        MODIFIER|WASH7P|ENSG00000227232|Transcript|ENST00000488147|\
        unprocessed_pseudogene||||||||||1|4373|-1||SNV|HGNC|HGNC:38034|YES||||||||\
        Ensembl|||||||||||||||,C|downstream_gene_variant|MODIFIER|WASH7P|653635|\
        Transcript|NR_024540.1|transcribed_pseudogene||||||||||1|4331|-1||SNV|\
        EntrezGene|HGNC:38034|YES||||||||RefSeq|||||||||||||||,C|\
        upstream_gene_variant|MODIFIER|DDX11L1|100287102|Transcript|NR_046018.2|\
        transcribed_pseudogene||||||||||1|1843|1||SNV|EntrezGene|HGNC:37102|YES\
        ||||||||RefSeq|||||||||||||||";
        let vep = Vep::from_str(s).unwrap();

        insta::assert_yaml_snapshot!(vep);
    }
}
