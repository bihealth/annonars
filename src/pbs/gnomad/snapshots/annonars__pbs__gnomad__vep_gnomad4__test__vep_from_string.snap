---
source: src/pbs/gnomad/vep_gnomad4.rs
expression: vep
---
allele: C
consequence: upstream_gene_variant
impact: MODIFIER
symbol: DDX11L1
gene: ENSG00000223972
featureType: Transcript
feature: ENST00000450305
featureBiotype: transcribed_unprocessed_pseudogene
alleleNum: 1
distance: "1979"
strand: "1"
variantClass: SNV
symbolSource: HGNC
hgncId: "HGNC:37102"
canonical: true
source: Ensembl

